text/0000-render-element-modifiers.md

```diff
@@ -0,0 +1,590 @@
+}
+```
+```hbs
+<div
+  {{did-insert (scroll-to) @scrollPosition}}
+  {{did-update (scroll-to) @scrollPosition}}
+  class="scroll-container"
+>
+  ...
+</div>
+```
+
+### Official Addon
+
+While these modifiers will be generally useful, modifiers are meant to be a more
+generic API that can be used to create libraries for solving specific problems.
+Unfortunately, the community hasn't had much time to experiment with modifiers,
+since the public API for them hasn't been finalized.
+
+The modifiers in this RFC will provide an basic stepping stone for users who
+want to emulate lifecycle hooks and incrementally convert their applications to
+modifiers while modifiers in general are being experimented with in the
+community. In time, users should be able to pick and choose the modifiers that
+suit their needs more directly and effectively, and they shouldn't have to
+include these modifiers in the payload. These modifiers should also not be seen
+as the "Ember way" - they are just another addon, a basic one supported by
+the Ember core team, but one which may or may not be appropriate for a given
+application.
+
+## Detailed design
+
+This RFC proposes adding three element modifiers:
+
+* `{{did-insert}}`
+* `{{did-update}}`
+* `{{will-destroy}}`
+
+Note that element modifiers do _not_ run in SSR mode - this code is only run on
+clients. Each of these modifiers receives a callback as it's first positional
+parameter:
+
+```ts
+type RenderModifierCallback = (element: Element, positionalArgs: [any], namedArgs: object): void;
+```
+
+The `element` argument is the element that the modifier is applied to,
+`positionalArgs` contains any remaining positional arguments passed to the
+modifier besides the callback, and `namedArgs` contains any named arguments
+passed to the modifier. If the first positional argument is not a callable
+function, the modifier will throw an error.
+
+> Note: The timing semantics in the following section were mostly defined in the
+> [element modifier manager RFC](https://github.com/emberjs/rfcs/blob/master/text/0373-Element-Modifier-Managers.md)
+> and are repeated here for clarity and convenience.
+
+### `{{did-insert}}`
+
+This modifier is activated only when The element is inserted in the DOM.
+
+It has the following timing semantics when activated:
+
+* **Always**
+  * called after DOM insertion
+  * called _after_ any child element's `{{did-insert}}` modifiers
+  * called _after_ the enclosing component's `willRender` hook
+  * called _before_ the enclosing component's `didRender` hook
+  * called in definition order in the template
+* **May or May Not**
+  * be called in the same tick as DOM insertion
+  * have the sibling nodes fully initialized in DOM
+
+Note that these statements do not refer to when the modifier is _activated_,
+only to when it will be run relative to other hooks and modifiers _should it be
+activated_. The modifier is only activated on insertion and arg changes.
```

Comment:
Pretty sure you want to remove `and arg changes` here, right?

Reply:
Ah yup, thanks for pointing that out!